core/app/models/spree/promotion.rb

```diff
@@ -91,23 +91,25 @@ def eligible_rules(promotable, options = {})
       # Promotions without rules are eligible by default.
       return [] if rules.none?
       eligible = lambda { |r| r.eligible?(promotable, options) }
-      specific_rules = rules.for(promotable)
+      specific_rules = rules.select { |rule| rule.applicable?(promotable) }
       return [] if specific_rules.none?
 
+      eligible_rules = specific_rules.map(&eligible)
+
       if match_all?
         # If there are rules for this promotion, but no rules for this
         # particular promotable, then the promotion is ineligible by default.
-        unless specific_rules.all?(&eligible)
+        unless eligible_rules.all?
           @eligibility_errors = specific_rules.map(&:eligibility_errors).detect(&:present?)
           return nil
         end
         specific_rules
       else
-        unless specific_rules.any?(&eligible)
+        unless eligible_rules.any?
           @eligibility_errors = specific_rules.map(&:eligibility_errors).detect(&:present?)
           return nil
         end
-        specific_rules.select(&eligible)
+        specific_rules.zip(eligible_rules).select {|rule, elligible| elligible}.map(&:first)
```

Comment:
Unused block argument - `rule`. If it's necessary, use `_` or `_rule` as an argument name to indicate that it won't be used.<br>Line is too long. [92/80]<br>Space between { and | missing.<br>Space missing inside }.